### PR TITLE
WIP: Add amount, card number, LocalDateTime...

### PR DESCRIPTION
**Pull request by Fleaman777**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -211,7 +211,7 @@
                         <phase>validate</phase>
                         <configuration>
                             <configLocation>google_checks.xml</configLocation>
-                            <violationSeverity>warning</violationSeverity>
+                            <!--<violationSeverity>warning</violationSeverity>-->
```
Comment (Meisssshu) [inline pom.xml]: Why? Wip?

File: src/main/java/com/epam/lab/paymentsystem/controller/AccountController.java [hunk 2]
```diff
@@ -82,4 +84,37 @@ public String addAccount(@ModelAttribute(name = "accountDto") AccountDto account
...
+   * @param accountId long
+   * @param amount    long
+   * @param model     Model
+   * @return String
+   */
+  @PostMapping(value = "/{userLogin}/account/{accountId}/addAmount")
+  public String addAmount(@PathVariable(name = "accountId") long accountId,
+                          @RequestParam(name = "amount") long amount,
+                          Model model) {
+    try {
```
Comment (Meisssshu) [inline src/main/java/com/epam/lab/paymentsystem/controller/AccountController.java]: Later on, you will have to refactor it using `ControllerAdvice`. Keep in mind

File: src/main/java/com/epam/lab/paymentsystem/entities/User.java [hunk 3]
```diff
@@ -24,7 +24,7 @@
   @Column(name = "login")
   private String login;
 
-  @ManyToOne(cascade = CascadeType.ALL)
+  @ManyToOne(cascade = CascadeType.REFRESH)
```
Comment (Meisssshu) [inline src/main/java/com/epam/lab/paymentsystem/entities/User.java]: Why to change?

File: src/main/java/com/epam/lab/paymentsystem/repository/CardRepository.java [hunk 4]
```diff
@@ -48,4 +51,18 @@
    * @return card entity
    */
   Card getCardById(long id);
+
+
+  /** Find card by number card.
+   * @param cardNumber String
+   * @return Card
+   */
+  Card findByCardNumber(String cardNumber);
```
Comment (Meisssshu) [inline src/main/java/com/epam/lab/paymentsystem/repository/CardRepository.java]: Why to have two similar methods - find and get?

File: src/main/java/com/epam/lab/paymentsystem/repository/CardRepository.java [hunk 5]
```diff
@@ -39,7 +42,7 @@
    * @param account account of card
    * @return list of cards by user
    */
-  List<Card> getAllCardsByAccountIsIn(List<Account> account);
+  //List<Card> getAllCardsByAccountIsIn(List<Account> account);
```
Comment (Meisssshu) [inline src/main/java/com/epam/lab/paymentsystem/repository/CardRepository.java]: If you don`t need it - fully remove, please

File: src/main/java/com/epam/lab/paymentsystem/repository/CardRepository.java [hunk 6]
```diff
@@ -25,6 +25,9 @@
    */
   List<Card> getAllByAccount(Account account);
 
+
+  List<Card> getAllByUserOrIsActive(User user, boolean active);
```
Comment (Meisssshu) [inline src/main/java/com/epam/lab/paymentsystem/repository/CardRepository.java]: I don't understand what is this method used for. The naming is too strange

File: src/main/java/com/epam/lab/paymentsystem/service/OperationService.java [hunk 7]
```diff
@@ -38,5 +39,7 @@
    * @param operationDto operation dto.
    */
   void makePayment(OperationDto operationDto);
+
+  String dateFormat(LocalDateTime date);
```
Comment (Meisssshu) [inline src/main/java/com/epam/lab/paymentsystem/service/OperationService.java]: Can you make this util method private in `OperationServiceImpl`?

File: src/main/java/com/epam/lab/paymentsystem/service/CardService.java [hunk 8]
```diff
@@ -28,9 +28,14 @@
...
 
+  /**
+   * Returns list of cards by current user.
+   * @return list of cards
+   */
   List<Card> getAllCardsByCurrentUser();
 
-  List<Card> getAllCardsByAccountIsIn(List<Account> account);
+
+  List<Card> getAllCardsByCurrentUserWithoutBlocked();
```
Comment (Meisssshu) [inline src/main/java/com/epam/lab/paymentsystem/service/CardService.java]: I think `getAllNonBlockedCardsOfCurrentUser` is the better naming

File: src/main/java/com/epam/lab/paymentsystem/service/impl/CardServiceImpl.java [hunk 9]
```diff
@@ -94,18 +95,56 @@ public Card createCard(CardDto card) throws UnsupportedOperationException {
     LOGGER.info("Creating new card");
     Account account = accountService.getAccountById(card.getAccountId());
     User user = userService.getUserByLogin(card.getUserLogin());
+    String cardNumber = generateCardNumber();
```
Comment (Meisssshu) [inline src/main/java/com/epam/lab/paymentsystem/service/impl/CardServiceImpl.java]: There is no need to create cardNumber if user is null, for example.
Please, put this method after the user check

File: src/main/java/com/epam/lab/paymentsystem/service/impl/CardServiceImpl.java [hunk 10]
```diff
@@ -94,18 +95,56 @@ public Card createCard(CardDto card) throws UnsupportedOperationException {
...
+  /**
+   * Return unique number card.
+   *
+   * @return string number card
+   */
+  private String generateCardNumber() {
+    String number = String.format("%04d", random.nextInt(10000));
+    Card cardWithNumber = cardRepository.findByCardNumber(number);
+    if (cardWithNumber != null) {
+      LOGGER.info("Number card already exist");
```
Comment (Meisssshu) [inline src/main/java/com/epam/lab/paymentsystem/service/impl/CardServiceImpl.java]: No need to log recursion. Better log the start and finish of recursion

File: src/main/java/com/epam/lab/paymentsystem/service/impl/OperationServiceImpl.java [hunk 11]
```diff
@@ -46,13 +47,25 @@
...
-    Card srcCard = cardService.getCardById(operationDto.getCardSrcId());
-    Card dstCard = cardService.getCardById(operationDto.getCardDstId());
+
+    Card srcCard = cardService.getCardByCardNumber(operationDto.getNumberSrcCard());
+    Card dstCard = cardService.getCardByCardNumber(operationDto.getNumberDstCard());
+
     Account srcAccount = srcCard.getAccount();
     Account dstAccount = dstCard.getAccount();
 
+    if (dstAccount.equals(null)) {
```
Comment (Meisssshu) [inline src/main/java/com/epam/lab/paymentsystem/service/impl/OperationServiceImpl.java]: Are you sure you want to use `equals`?

File: src/main/java/com/epam/lab/paymentsystem/service/impl/OperationServiceImpl.java [hunk 12]
```diff
@@ -46,13 +47,25 @@
...
+      LOGGER.error("This card not exist");
+      throw new UnsupportedOperationException("This card not exist");
+    }
+    LocalDateTime enter = LocalDateTime.now().withNano(0);
+    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
```
Comment (Meisssshu) [inline src/main/java/com/epam/lab/paymentsystem/service/impl/OperationServiceImpl.java]: You may put it to constants

File: src/main/java/com/epam/lab/paymentsystem/service/impl/OperationServiceImpl.java [hunk 13]
```diff
@@ -46,13 +47,25 @@
...
+    String formattedDateTime = enter.format(formatter);
+    enter = LocalDateTime.parse(formattedDateTime, formatter);
```
Comment (Meisssshu) [inline src/main/java/com/epam/lab/paymentsystem/service/impl/OperationServiceImpl.java]: I don't understand the purpose of second formatting using `LocalDateTime`

File: src/test/java/com/epam/lab/paymentsystem/service/CardServiceTest.java [hunk 14]
```diff
@@ -14,78 +20,73 @@
...
-    public void testGetAllCardsByAccountIdReturnsCardsList() {
-        when(accountService.getAccountById(accountId)).thenReturn(account);
-        when(cardRepository.getAllByAccount(account)).thenReturn(cards);
-        assertEquals(cards, cardService.getAllCardsByAccountId(accountId));
-    }
+  @Test
+  public void testGetAllCardsByAccountIdReturnsCardsList() {
+    when(accountService.getAccountById(accountId)).thenReturn(account);
+    when(cardRepository.getAllByAccount(account)).thenReturn(cards);
+    assertEquals(cards, cardService.getAllCardsByAccountId(accountId));
```
Comment (Meisssshu) [inline src/test/java/com/epam/lab/paymentsystem/service/CardServiceTest.java]: Please, add message and reformat code

File: src/test/java/com/epam/lab/paymentsystem/service/CardServiceTest.java [hunk 15]
```diff
@@ -14,78 +20,73 @@
...
-    public void testGetAllCardsByUserLoginReturnsCardsList() {
-        when(userService.getUserByLogin(user.getLogin())).thenReturn(user);
-        when(cardRepository.getAllByUser(user)).thenReturn(cards);
-        assertEquals(cards, cardService.getAllCardsByLogin(login));
-    }
+  @Test
+  public void testGetAllCardsByUserLoginReturnsCardsList() {
+    when(userService.getUserByLogin(user.getLogin())).thenReturn(user);
+    when(cardRepository.getAllByUser(user)).thenReturn(cards);
+    assertEquals(cards, cardService.getAllCardsByLogin(login));
```
Comment (Meisssshu) [inline src/test/java/com/epam/lab/paymentsystem/service/CardServiceTest.java]: Please, add message and reformat code

Comment (Fleaman777) [reply src/test/java/com/epam/lab/paymentsystem/service/CardServiceTest.java]: Do I need to write a new test code?

File: src/main/java/com/epam/lab/paymentsystem/entities/User.java (hunk 3 above)
Comment (Fleaman777) [reply src/main/java/com/epam/lab/paymentsystem/entities/User.java]: We decided that you can use this type, the main thing that would be the same everywhere

File: src/test/java/com/epam/lab/paymentsystem/service/CardServiceTest.java (hunk 15 above)
Comment (Meisssshu) [reply src/test/java/com/epam/lab/paymentsystem/service/CardServiceTest.java]: No, you have to fix it to the next way:
```
assertEquals(
   "Test should return cards of given user!",
    cards,
    cardService.getAllCardsByLogin(login)
);
```

File: src/main/java/com/epam/lab/paymentsystem/controller/AccountController.java [hunk 16]
```diff
@@ -25,20 +27,39 @@
...
+   * @param servletRequest servlet request
+   * @return redirect to previous page
+   */
   @PostMapping(value = "/{userLogin}/account/{accountId}/block")
-  public String blockAccount(@PathVariable(name = "accountId") long id) {
+  public String blockAccount(@PathVariable(name = "accountId") long id,
+                             HttpServletRequest servletRequest) {
     accountService.blockAccountById(id);
-    return REDIRECT_TO + "/{userLogin}";
+    String referer = servletRequest.getHeader("Referer");
```
Comment (Meisssshu) [inline src/main/java/com/epam/lab/paymentsystem/controller/AccountController.java]: Is it used to stay on page?

File: src/test/java/com/epam/lab/paymentsystem/service/CardServiceTest.java [hunk 17]
```diff
@@ -56,37 +69,44 @@ public void startUp() {
...
+        cardService.getAllCardsByAccountId(accountId, pageable),
+        "The card page should be equal to the card page retrieved by the card service");
   }
 
   @Test
   public void testGetAllCardsByUserLoginReturnsCardsList() {
     when(userService.getUserByLogin(user.getLogin())).thenReturn(user);
     when(cardRepository.getAllByUser(user)).thenReturn(cards);
-    assertEquals(cards, cardService.getAllCardsByLogin(login));
+    assertEquals(cards,
```
Comment (Meisssshu) [inline src/test/java/com/epam/lab/paymentsystem/service/CardServiceTest.java]: Please, reformat code

File: src/test/java/com/epam/lab/paymentsystem/service/CardServiceTest.java [hunk 18]
```diff
@@ -56,37 +69,44 @@ public void startUp() {
...
-//    when(userService.getUserByLogin(user.getLogin())).thenReturn(user);
-//    when(cardRepository.save(card)).thenReturn(card);
-//    assertEquals(card, cardService.createCard(cardDto));
-//  }
+  @Test
+  public void testCreateCardReturnsCard() {
+    when(accountService.getAccountById(accountId)).thenReturn(account);
+    when(userService.getUserByLogin(user.getLogin())).thenReturn(user);
+    when(cardRepository.save(card)).thenReturn(card);
+    assertEquals(card, cardService.createCard(cardDto));
```
Comment (Meisssshu) [inline src/test/java/com/epam/lab/paymentsystem/service/CardServiceTest.java]: Please, add message and reformat code

File: src/main/java/com/epam/lab/paymentsystem/controller/AccountController.java (hunk 16 above)
Comment (Fleaman777) [reply src/main/java/com/epam/lab/paymentsystem/controller/AccountController.java]: Yes

**Pull request merged**